datafusion/expr-common/src/columnar_value.rs

```diff
@@ -89,7 +89,94 @@ pub enum ColumnarValue {
     /// Array of values
     Array(ArrayRef),
     /// A single value
-    Scalar(ScalarValue),
+    Scalar(Scalar),
+}
+
+#[derive(Clone, Debug)]
+pub struct Scalar {
+    value: ScalarValue,
+    data_type: DataType,
+}
```

Comment:
This is the main change of this PR

Reply:
I was thinking that I could open a new PR introducing just introducing this columnar value type, and then little by little removing the scalar value variants (Dictionary, Large* and *View...)

Reply:
I agree with @findepi 's suggestion to make the smallest PR as possible to begin. I think splitting this PR out into a small part to begin, along with comments explaining the planned end state (or linking back to the ticket that describes this) would be very helpful

Reply:
I agree with @findepi 's suggestion to make the smallest PR as possible to begin. I think splitting this PR out into a small part to begin, along with comments explaining the planned end state (or linking back to the ticket that describes this) would be very helpful